File: Project_4/BT13/cpp/dfs_general.cpp
#include <iostream>
#include <vector>
#include <stack>

// Depth-first search algorithm for a general graph (repeated edges and loops allowed)
// Author: Nguyễn Ngọc Thạch
// This program performs DFS traversal on a general graph starting from a given vertex.

using namespace std;

// Recursive DFS for general graph
void dfs_recursive(const vector<vector<int>>& adj, vector<bool>& visited, int vertex) {
    visited[vertex] = true;
    cout << vertex << " ";
    for (int neighbor : adj[vertex]) {
        if (!visited[neighbor]) {
            dfs_recursive(adj, visited, neighbor);
        }
    }
}

// Iterative DFS for general graph
void dfs_iterative(const vector<vector<int>>& adj, int start) {
    int n = adj.size();
    vector<bool> visited(n, false);
    stack<int> s;
    s.push(start);
    while (!s.empty()) {
        int vertex = s.top();
        s.pop();
        if (!visited[vertex]) {
            visited[vertex] = true;
            cout << vertex << " ";
            for (int i = adj[vertex].size() - 1; i >= 0; --i) {
                int neighbor = adj[vertex][i];
                if (!visited[neighbor]) {
                    s.push(neighbor);
                }
            }
        }
    }
}

int main() {
    int n, m;
    cout << "Enter number of vertices and edges: ";
    cin >> n >> m;
    vector<vector<int>> adj(n);
    cout << "Enter edges (u v) for each edge (0-indexed vertices, loops and repeated edges allowed):\n";
    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u); // Undirected general graph: add both directions, including loops
    }
    int start;
    cout << "Enter starting vertex: ";
    cin >> start;
    cout << "\nDFS using recursion: ";
    vector<bool> visited(n, false);
    dfs_recursive(adj, visited, start);
    cout << endl;
    cout << "DFS using iteration: ";
    dfs_iterative(adj, start);
    cout << endl;
    return 0;
}

File: Project_4/BT13/py/dfs_general.py
# Depth-first search algorithm for a general graph (repeated edges and loops allowed)
# Author: Nguyễn Ngọc Thạch
# This program performs DFS traversal on a general graph starting from a given vertex.

# Recursive DFS for general graph
def dfs_recursive(adj, visited, vertex):
    visited[vertex] = True
    print(vertex, end=" ")
    for neighbor in adj[vertex]:
        if not visited[neighbor]:
            dfs_recursive(adj, visited, neighbor)

# Iterative DFS for general graph
def dfs_iterative(adj, start):
    n = len(adj)
    visited = [False] * n
    s = []
    s.append(start)
    while s:
        vertex = s.pop()
        if not visited[vertex]:
            visited[vertex] = True
            print(vertex, end=" ")
            # Push neighbors in reverse to maintain DFS order
            for neighbor in reversed(adj[vertex]):
                if not visited[neighbor]:
                    s.append(neighbor)

def main():
    n, m = map(int, input("Enter number of vertices and edges: ").split())
    adj = [[] for _ in range(n)]
    print("Enter edges (u v) for each edge (0-indexed vertices, loops and repeated edges allowed):")
    for _ in range(m):
        u, v = map(int, input().split())
        adj[u].append(v)
        adj[v].append(u)  # Undirected general graph: add both directions, including loops

    start = int(input("Enter starting vertex: "))

    print("\nDFS using recursion: ", end="")
    visited = [False] * n
    dfs_recursive(adj, visited, start)
    print()

    print("DFS using iteration: ", end="")
    dfs_iterative(adj, start)
    print()

if __name__ == "__main__":
    main()
